lib/shopify_transporter/pipeline/magento/order/discounts.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+require 'shopify_transporter/pipeline/stage'
+require 'shopify_transporter/shopify'
+
+module ShopifyTransporter
+  module Pipeline
+    module Magento
+      module Order
+        class Discounts < Pipeline::Stage
+          def convert(input, record)
+            record.merge!(
+              {
+                discounts: discounts(input),
+              }.stringify_keys
+            )
+          end
+
+          private
+
+          def discounts(hash)
+            [
+              shipping_discount(hash),
+              fixed_amount_discount(hash),
+              percentage_discount(hash),
+            ].compact
+          end
+
+          def shipping_discount(hash)
+            shipping_discount_amount = value_as_float(hash, 'shipping_discount_amount')
+            shipping_amount = value_as_float(hash, 'shipping_amount')
+            if shipping_discount_amount > shipping_amount
+              {
+                code: discount_code(hash),
+                amount: shipping_discount_amount,
+                type: 'shipping',
+              }.stringify_keys
+            end
+          end
+
+          def fixed_amount_discount(hash)
+            discount_amount = discount_amount(hash)
+            discount_percentage = discount_percentage(hash)
+            if discount_amount > 0 && discount_percentage == 0
+              {
+                code: discount_code(hash),
+                amount: discount_amount,
+                type: 'fixed_amount',
+              }.stringify_keys
+            end
+          end
+
+          def percentage_discount(hash)
+            discount_amount = discount_amount(hash)
+            discount_percentage = discount_percentage(hash)
+
+            if discount_amount > 0 && discount_percentage != 0
+              {
+                code: discount_code(hash),
+                amount: discount_percentage,
+                type: 'percentage',
+              }.stringify_keys
+            end
+          end
+
+          def discount_percentage(hash)
+            return 0 unless hash.dig('items', 'result', 'items', 'item').present?
+
+            if line_items(hash).is_a?(Hash)
+              value_as_float(line_items(hash), 'discount_percent')
+            else
+              discounts = line_items(hash).map do |line_item|
+                value_as_float(line_item, 'discount_percent')
+              end.uniq
+              discount_applied_on_all_line_items?(discounts) ? discounts.first : 0
+            end
```

Comment:
this would be fine, i think the current implementation is fine too though -- especially if the if/elsif construct was swapped for a case like so:

```
case line_items(hash)
when Hash
...
when Array
...
end
```

this works without the need for any `is_a?` calls, because case/when constructs automatically match using `===`, and `Hash === {}`.

(unintuitively, though, `{} === Hash` is not true. the `===` operator is not symmetric for some reason.

```
irb(main):003:0> Hash === {}
=> true
irb(main):004:0> Hash == {}
=> false
```

ask me about this if you're curious, it's one of the things that annoys me about Ruby